Renamed newsletter_id and email_recipient_filter options

SimonBackx opened the pull request:
refs https://github.com/TryGhost/Team/issues/1596

- Renamed `newsletter_id` to `newsletter` option, the `newsletter` option expects a slug instead of an id.
- Renamed `email_recipient_filter` to `email_segment` option
- Default `email_segment` to `all`. Ignored if no newsletter is set
- `email_segment` is ignored if no newsletter is set
- When reverting a post to a draft, both `newsletter` and `email_segment` are reset to their default values (null, all)
- Removed legacy mapping from old email_recipient_filter values 'paid' and 'free' (already a migration in place)
- Dropped legacy throwing errors when email_recipient_filter is paid or free in transformEmailRecipientFilter
- Reorganized transformEmailRecipientFilter parameters for the now required newsletter parameter
- Fixed an issue where the newsletter filter wasn't working because it wasn't in permittedoptions
- Fixed an issue where you could send to an archived newsletter
- Added an extra protection when scheduling to an active, and later archiving the newsletter
- Dropped support for `send_email_when_published` in API
- When importing posts we currently don't have a system in place to set the newsletter_id to map the `send_email_when_published` behaviour. Since this was already the case, I won't include a fix in this PR.
- Stripped `email_recipient_filter`/`email_segment` from Content API (https://ghost.slack.com/archives/C02G9E68C/p1652363211841359?thread_ts=1650623650.233229&cid=C02G9E68C)
- Updated `admin-api-schema` to 3.2.0, which includes the new email_segment property
- Contains a temporary fix for https://github.com/TryGhost/Team/issues/1626, where the `.related('newsletter').fetch` call fails when the newsletter relation is already loaded, because of the overridden `formatOnWrite` method.

Since the `email_recipient_filter` is no longer used without a newsletter, the `none` value is no longer used. A migration transforms all those values to `all`. This should be safe, because we only send an email now when newsletter_id is not null (scheduled posts should already have a newsletter_id, even if at the time of scheduling they didn't add the newsletter_id option, because at that time, we defaulted to the default newsletter).

Admin changes to make this work: https://github.com/TryGhost/Admin/pull/2380 (will get merged at the same time)

Commit:
Renamed newsletter_id and email_recipient_filter options

refs https://github.com/TryGhost/Team/issues/1596

- Renamed `newsletter_id` to `newsletter` option
- Renamed `email_recipient_filter` to `email_segment` option
- Default `email_segment` to `all`. Ignored if no newsletter is set
- `email_segment` is ignored if no newsletter is set
- When reverting a post to a draft, both `newsletter` and `email_segment` are reset to their default values (null, all)
- Removed legacy mapping from old email_recipient_filter values 'paid' and 'free' (already a migration in place)
- Dropped legacy throwing errors when email_recipient_filter is paid or free in transformEmailRecipientFilter
- Reorganized transformEmailRecipientFilter parameters for the now required newsletter parameter
- Fixed an issue where the newsletter filter wasn't working because it wasn't in permittedoptions
- Fixed an issue where you could send to an archived newsletter
- Added an extra protection when scheduling to an active, and later archiving the newsletter
- Dropped support for `send_email_when_published` in API
- When importing posts we currently don't have a system in place to set the newsletter_id to map the `send_email_when_published` behaviour. Since this was already the case, I won't include a fix in this PR.
- Stripped `email_recipient_filter`/`email_segment` from Content API (https://ghost.slack.com/archives/C02G9E68C/p1652363211841359?thread_ts=1650623650.233229&cid=C02G9E68C)
- Updated `admin-api-schema` to 3.2.0, which includes the new email_segment property

SimonBackx commented on core/server/api/canary/utils/serializers/output/mappers/posts.js:
```diff
@@ -57,8 +61,6 @@ module.exports = async (model, frame, options = {}) => {
         date.forPost(jsonModel);
         gating.forPost(jsonModel, frame);
-
-        delete jsonModel.newsletter_id;
```
Note that this is dropped because this is removed a couple of lines later in `clean.post(jsonModel, frame);` in both the content and admin api

matthanley commented on core/server/services/posts/posts-service.js:
```diff
@@ -17,69 +17,35 @@ class PostsService {
-
-        // Make sure the newsletter_id is matching an active newsletter
-        if (frame.options.newsletter_id) {
-            const newsletter = await this.models.Newsletter.findOne({id: frame.options.newsletter_id, filter: 'status:active'}, {transacting: frame.options.transacting});
```
Did this ever work? 😬

SimonBackx replied:
No, we didn't have tests for this 😬 We even tested with archived newsletters without knowing 😅

Commit:
Added hack for newsletter fetching

refs https://github.com/TryGhost/Team/issues/1626
refs https://github.com/TryGhost/Team/issues/1596

Some error causes bookshelf not to be able to load the newsletter relation when it is already loaded.

SimonBackx commented on core/server/models/post.js:
```diff
@@ -675,28 +659,30 @@ Post = ghostBookshelf.Model.extend({
+            // email_segment is read-only and should only be set using a query param when publishing/scheduling
+            // we can't set it if we don't pass newsletter
+            if (options.email_segment
+                && (options.email_segment !== 'none')) {
```
I've added a migration to drop that value.

SimonBackx commented:
Note: tests fail again because `@tryghost/admin-api-schema` reverted a required change in 3.2.1

matthanley commented on core/server/data/migrations/versions/5.0/2022-05-13-11-38-drop-none-email-recipient-filter.js:
```diff
@@ -0,0 +1,25 @@
+        logging.info(`Updating posts with email_recipient_filter 'none' to 'all'`);
+
+        const affectedRows = await knex('posts')
+            .update('email_recipient_filter', 'all')
```
Is it worth also explicitly setting `newsletter_id:null` here?

matthanley commented on core/server/services/mega/mega.js:
```diff
@@ -213,10 +196,21 @@ const addEmail = async (postModel, options) => {
 
+    if (newsletter.get('status') !== 'active') {
+        // A post might have been scheduled to an archived newsletter.
+        // Don't send it (people can't unsubscribe any longer).
```
What happens to the post in this situation? I assume retrying the send would fail again, but is there a path to update the newsletter and resend?

SimonBackx replied:
Well, no email will get attached to the post. So they can retry by reverting to draft and publishing again with a different newsletter

matthanley commented on core/server/services/mega/mega.js:
```diff
@@ -423,11 +417,8 @@ async function getEmailMemberRows({emailModel, memberSegment, options}) {
-        newsletter = await emailModel.related('newsletter').fetch(Object.assign({}, {require: false}, _.pick(options, ['transacting'])));
-    }
-    const recipientFilter = transformEmailRecipientFilter(emailModel.get('recipient_filter'), {errorProperty: 'recipient_filter'}, newsletter);
+    const newsletter = await emailModel.related('newsletter').fetch(Object.assign({}, {require: true}, _.pick(options, ['transacting'])));
```
Does this need the same hack as above?

SimonBackx replied:
Not that I noticed. Main problem is that I cannot reproduce it in tests 😬

SimonBackx replied:
I'll add it to be sure + gonna continue to investigate the issue

Commit:
Updated migration to also set newsletter_id to null

Commit:
Fixed formatOnWrite issue and updated fixtures

refs https://github.com/TryGhost/Team/issues/1626

Updated newsletter fixtures to be able to reproduce the issue in all tests.

SimonBackx commented on core/server/services/mega/mega.js:
```diff
@@ -53,8 +53,9 @@ const getReplyToAddress = (fromAddress, replyAddressOption) => {
  * @param {Object} options
  */
 const getEmailData = async (postModel, options) => {
-    let newsletter = await postModel.related('newsletter').fetch();
+    let newsletter = postModel.relations.newsletter ?? await postModel.related('newsletter').fetch();
     if (!newsletter) {
+        // The postModel doesn't have a newsletter in test emails
         newsletter = await models.Newsletter.getDefaultNewsletter();
```
Permalinking this change for reference

SimonBackx commented on core/server/services/mega/mega.js:
```diff
@@ -53,8 +53,9 @@ const getReplyToAddress = (fromAddress, replyAddressOption) => {
  */
 const getEmailData = async (postModel, options) => {
-    let newsletter = await postModel.related('newsletter').fetch();
+    let newsletter = postModel.relations.newsletter ?? await postModel.related('newsletter').fetch();
     if (!newsletter) {
+        // The postModel doesn't have a newsletter in test emails
```
Permalinking this change

Commit:
Switched newsletter id to newsletter slug in the options

refs https://ghost.slack.com/archives/C02G9E68C/p1652459562417399

- A slug makes more sense as an option in the API than the id.
- Requires some changes to the admin
